File: GUILib/src/space/kiibou/gui/ImageBuffer.java
package space.kiibou.gui;

import processing.core.PImage;

import javax.imageio.ImageIO;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Objects;

public class ImageBuffer {

    private static final HashMap<String, PImage> buffer;

    static {
        buffer = new HashMap<>();
    }

    public static PImage loadImage(String path) {
        return buffer.computeIfAbsent(path, p -> {
            try {
                InputStream stream = GraphicsElement.class.getClassLoader().getResourceAsStream(path);
                Objects.requireNonNull(stream);
                return new PImage(ImageIO.read(stream));
            } catch (Exception e) {
                System.out.printf("Error loading image \"%s\"%n", path);
                return new PImage(0, 0);
            }
        });
    }

}

File: src/guiLibMain/kotlin/space/kiibou/gui/ImageBuffer.kt
package space.kiibou.gui

import processing.awt.PImageAWT
import processing.core.PImage
import javax.imageio.ImageIO

private var buffer: HashMap<String, PImage> = HashMap()

fun loadImage(path: String): PImage {
    return buffer.computeIfAbsent(path) {
        try {
            val stream = GraphicsElement::class.java.classLoader.getResourceAsStream(path)!!
            PImageAWT(ImageIO.read(stream))
        } catch (e: Exception) {
            println("Error loading image <$path>")
            PImage(-1, -1)
        }
    }
}
